lib/curlable.js

```diff
@@ -5,182 +5,320 @@ const bodyParser = require('body-parser');
+    if (terminationCallback) {
+      terminationCallback();
+    }
+  }
+
+  _onSpawnedProcessLine(line) {
+    if (!this.spawnedProcess || !this.ready) {
+      if (this.prompt) {
+        if (line === this.prompt) {
+          this.ready = true;
+          this.logger('Ready after ' + (Date.now() - this.startTime) + 'ms.\n');
+        } else {
+          // Ignore spawned process output before prompt.
+        }
+      } else {
+        // Ignore spawned process output after kill.
+      }
+    } else if (this.currentQueueItem) {
+      const queueItem = this.currentQueueItem;
+
+      // Accumulate output lines except the prompt, if configured.
+      if (!this.prompt || line !== this.prompt) {
+        if (queueItem.accumulatedOutput) {
+          queueItem.accumulatedOutput += line + '\n';
+        } else {
+          queueItem.accumulatedOutput = line + '\n';
+        }
       }
-    };
 
-    if (this.options && this.options.prompt) {
-      if (line === this.options.prompt) {
-        serveResult(this.accumulatedOutput);
+      if (!this.prompt || line === this.prompt) {
+        this.currentQueueItem = null;
+        this._finishQueueItem(queueItem);
       } else {
-        this.accumulatedOutput += line + '\n';
+        // Wait for prompt.
       }
     } else {
-      serveResult(line + '\n');
+      const error = new Error('The tool is generating output when it is not expected:\n' + line);
+      const errorCallback = this.errorCallback;
+      if (errorCallback) {
+        errorCallback(error);
+      }
     }
-  };
-  this.accumulatedOutput = '';
-  this.spawnedProcess.stdin.write(this.queriesQueue.query + '\n');
-};
+  }
 
-Curlable.prototype.runQuery = function(query, callback) {
-  var lines = query.split('\n');
-  var realQuery = [];
-  for (var i = 0; i < lines.length; ++i) {
-    lines[i] = lines[i].trim();
-    if (lines[i].length) {
-      realQuery.push(lines[i]);
+  _processQueue() {
+    // Ignore the call if killed, currently processing, or no more items.
+    if (!this.spawnedProcess || this.currentQueueItem || !this.queriesQueue) {
+      return;
     }
+
+    this.currentQueueItem = this.queriesQueue;
+    this.queriesQueue = this.queriesQueue.next;
+
+    this.currentQueueItem.writeTimestamp = Date.now();
+    this.logger(this.currentQueueItem.id + '\t-\t' + JSON.stringify(this.currentQueueItem.query) + '\n');
+    this.spawnedProcess.stdin.write(this.currentQueueItem.query + '\n');
   }
-  if (realQuery.length === 0) {
-    callback('Need a nonempty query.\n');
-    return;
-  }
-  if (realQuery.length !== 1) {
-    callback('Need a single-line query.\n');
-    return;
-  }
-  query = realQuery[0];
-  if (!this.queriesQueue) {
-    // No open queue. Start one, and kick off `runQueries()`.
-    this.queriesQueue = {
-      query,
-      callback,
+
+  /**
+   * Queues a query to be processed by the underlying tool.
+   *
+   * @param  {string}   query     The query input.
```

Comment:
The number of spaces looks inconsistent to me here.

Reply:
True.